### ISSUE TITLE
Error when adding a dashboard from the marketplace: «Illegal character in path at index 0» 

### ISSUE THREAD

**Issue by heyanctil** [ozoneplatform/owf-framework#70]
We are using the latest release of OWF and we have start playing with the Marketplace. We have encountered this week a problem when we tried to add a Dashboard:

«Illegal character in path at index 0: \\marketplace/public/...

![marketplace_error](https://cloud.githubusercontent.com/assets/7291291/12202067/c6a0ab64-b5f5-11e5-9265-1b22c5274f6a.png)

While doing our investigation, we have found in the apache-tomcat forlder, a file (**prodDB.script**) with that line:

    INSERT INTO WIDGET_DEFINITION VALUES(196,1,'LCSS',200,'/marketplace/themes/gold.theme/images/Mp_logo.png','/marketplace/themes/gold.theme/images/Mp_logo.png',FALSE,TRUE,'5673d152-7504-994e-53b5-9ff343710c8f','\marketplace','null',200,FALSE,NULL,'\marketplace/public/storeDescriptor',NULL,FALSE)

**I don't know if it's related to our problem but we don't understand the used of backslash in the url and clearly this file is used by OWF!**

We've also found while digging in the code that maybe the problem could be in this service:

    class MarketplaceService extends BaseService {
        ...
        def addExternalStackToUser(params) {
            def stMarketplaceJson = new HashSet()
            stMarketplaceJson.addAll(buildWidgetListFromMarketplace(params.guid))

            if (!stMarketplaceJson.isEmpty()) {
                def listings = addListingsToDatabase(stMarketplaceJson)
                if (listings.size() == 1) {
                    stackService.addToUser(listings.get(0), accountService.getLoggedInUser())
                }
            }
            else {
                throw new OwfException(message:'No data received from Marketplace',
                    exceptionType: OwfExceptionTypes.GeneralServerError)
            }
        }

I'm not familiar with Grails or Groovy but I tried to install it to build OWF, but working in an offline environment complicate our task to find a solution!

**Comment by rpokorny** [comment]
I agree that the URLs in that widget definition appear to be wrong.  The default data for OWF does not include any marketplace widgets however, so I believe that widget definition must have been added by you or someone else in your group.  You should be able to use the App Components Manager admin widget to update the URLs of that widget definition.

**Comment by savardmario** [comment]
In order to test the Marketplace interaction with OWF, I have done the following steps in order to attempt to share a Dashboard that has been publish on the Marketplace.

I have used the following packages:

OWF-bundle-7.16.1-GA.zip
marketplace-bundle-7.16.2.zip

I have used 2 differents VM on the same network to host OWF and Marketplace on different machine:

VM_MP:  https://vm_mp:8443/marketplace
VM_OWF: https://vm_owf:8443/owf/

I have also created a second administrator account called testAdmin2 in /lib/users.properties (copy testAdmin1 line and change testAdmin1 for testAdmin2) in both OWF and Marketplace.

I have used JAVA 8:
JAVA: jdk1.8.0_60

Here are the steps:

1. Extract OWF on vm_owf. Launch OWF. 
   Ensure it is up and running and logging using testAdmin1.
   Browse https://vm_owf:8443/owf/ to ensure it shows up.
2. Extract MP on vm_mp. Launch MP. 
   Ensure it is up and running by logging using testAdmin1.
   Browse https://vm_mp:8443/marketplace to ensure it shows up.
3. Browse OWF. Reach Administration panel. 
   Add a Store using https://vm_mp:8443/marketplace URL.
   Press F5 and ensure the Store is available through the OWF toolbar.
4. Create a new empty dasboard called Test1 selecting a random pre-defined template.
   Go back to the App selector.
   Choose the small "i" icon and use "Publish to Store".
   From the Store, follow the submit process. Use Submit, Outside, Approve buttons. 
   Click Start to launch the dashboard.
5. From another browser, log using testAdmin2 on https://vm_owf:8443/owf/
   Reach the Store. Choose the Test1 dashboard and click "Add".
   
   At this point, I get an error message saying:
   Error {status: 500, responseText: "{"success":false,"errorMsg":" No data received from Marketplace"}", stack: (...), message: "Unable to load /owf/stack/ status:500", $className: "Ext.Base"…}message: "Unable to load /owf/stack/ status:500"responseText: "{"success":false,"errorMsg":" No data received from Marketplace"}"
   
The full stack in OWF is in attachment:

Basically, I managed to configure OWF to use a Marketplace, I managed to publish an empty dashboard and pass through the approval process workflow to make it available to other user.
What I didn't manage to do is to get this new dashboard in OWF for another user (does not event work if I use the same user testAdmin1).

Is there anything missing to make this process work? I have tried to stay as "vanilla" as possible. I'm I missing something?

Help on this topic would be greatly appreciated.

Thanks
[owf-stacktrace.txt](https://github.com/ozoneplatform/owf-framework/files/96160/owf-stacktrace.txt)

**Comment by mparizernc** [comment]
Please have a look at the  OWF Configuration Guide at Section:4.2.1.4 Store Synchronization you may be missing items in your security context files that are preventing services from talking to each other.

**Comment by savardmario** [comment]
I tried to do the same experience while hosting both OWF and OMP on the same machine by configuring different ports. The experiment allowed me to share Dashboard correctly without error. I'll try to figure out why it is not working in our environment (not local). I will look into security context.

Also, I still have to figure out how .war file get synchronized from the marketplace into the OWF instance since launch URL is only pointing to a .html file and not a .war file.

Thanks for the hint.

**Comment by mparizernc** [comment]
I am not sure what you mean by "how .war file get synchronized from the marketplace into the OWF instance since launch URL is only pointing to a .html file and not a .war file." ?

Marketplace will send a rest call to OWF with information about a listing/dashboard. OWF will then create the information within OWF.  OWF and Marketplace are separate programs that appear to work as one program through the use of REST calls to create similar data in each program.

**Comment by savardmario** [comment]
I'm wondering where the .war file (widget deployment package) should be located: marketplace only, marketplace and OWF or only OWF? At first, I would think the Marketplace would be the placeholder for .war file. Than, those .war would be copied over to an OWF instance such that the server would host them locally. But maybe this is not the deployment scenario. I would expect the OWF instance to be independent of the Marketplace. Can you clarify the workflow related to .war file?

Thanks

**Comment by mparizernc** [comment]
I believe I understand now. Neither Marketplace or OWF actually hold/host a widgets .war file.  They both contain a launch URL which point to where ever the widget is installed (which would be a server separate from OWF and Marketplace). When you add a widget from marketplace to OWF, marketplace sends the launch url and other metadata to OWF.  OWF when it launches the widget uses the launch url to open the widget in an Iframe.  OWF is just running the widget from the external server where the widget is installed.  I hope this makes more sense.

**Comment by savardmario** [comment]
Yes it does. However, since the beginning, we have been dropping our widget in the OWF webapps folder and expected the Marketplace to be the place to put the widget eventually, almost like Apple Store or Google Play does. 

From what I understand now, these widgets could be hosted in a central location somewhere. I though this central location would be the marketplace and a local copy would occur in OWF. In your explanation, it means many OWF instances will do remote call to a unique server. So OWFs would depend on that specific location to be up and running while hosting directly in OWF makes the OWF instance stand alone. 
I was expecting a behavior nearer what we are used to with Phones and Tablet which makes the devices able to run the apps locally once fetched from the store.

**Comment by mparizernc** [comment]
I am afraid at this time this functionality is not available.  Most of the use cases for Marketplace and OWF from the community are for applications spread across multiple locations.  In most cases the team administrating OWF and Marketplace do not have access to the actual widget software to make it downloadable.

**Comment by savardmario** [comment]
Just to close on this (because we are getting far from the original topic), I don't see much how the Marketplace covers much more than OWF since OWF was already able to "mash-up" web applications spread across multiple location using the URL in descriptors. I understand it is a central location to managing and exposing content from many locations in a fashionably manner (rating, comments, stars, etc). Thanks for those clarifications.
